skoolkit/basic.py

```diff
@@ -114,14 +115,92 @@
     255: 'COPY'
 }
 
+def get_number(snapshot, i):
+    if snapshot[i] == 0:
+        # Small integer (signed)
+        num = _get_integer(snapshot, i)
+    else:
+        # Floating point number (signed)
+        num = _get_float(snapshot, i)
+    return num
+
+def _get_integer(snapshot, i):
+    # See ZX Spectrum ROM routine INT_FETCH at 0x2D7F
+    sign = snapshot[i + 1]
+    lsb = (snapshot[i + 2] ^ sign) - sign
+    if lsb < 0:
+        lsb += 256
+        carry = 1
+    else:
+        carry = 0
+
+    msb = ((snapshot[i + 3] + sign + carry) & 255) ^ sign
+
+    num = 256 * msb + lsb
+    if (sign & 128) != 0:
+        num *= -1
+
+    return num
+
+def _get_float(snapshot, i):
+    exponent = snapshot[i] - 160
+    if (snapshot[i + 1] & 128) == 0:
+        sign = 1
+    else:
+        sign = -1
+    mantissa = float(16777216 * (snapshot[i + 1] | 128)
+                     + 65536 * snapshot[i + 2]
+                     + 256 * snapshot[i + 3]
+                     + snapshot[i + 4])
+    num = sign * mantissa * (2 ** exponent)
+    return num
+
+class TextReader:
+    def __init__(self):
+        self.lspace = False
+
+    def get_chars(self, code):
+        if code <= 32:
+            self.lspace = False
+            return self._get_char(code)
+        if code <= 164:
+            self.lspace = True
+            return self._get_char(code)
+        return self._get_token(code)
+
+    def _get_char(self, code):
+        if code == 94:
+            return '↑'
+        if code == 96:
+            return '£'
+        if code == 127:
+            return '©'
+        if 32 <= code <= 126:
+            return chr(code)
+        return '{{0x{:02X}}}'.format(code)
+
+    def _get_token(self, code):
+        token = TOKENS[code]
+        if self.lspace and code >= 197 and token[0] >= 'A':
+            token = ' ' + token
+        self.lspace = True
+        if code < 168 or code == 203 or token[-1] in '#=>':
+            # RND, INKEY$, PI, THEN, '<=', '>=', '<>', 'OPEN #', 'CLOSE #'
+            return token
+        self.lspace = False
+        return token + ' '
+
 class BasicLister:
+    def __init__(self):
+        self.text = TextReader()
+
     def list_basic(self, snapshot):
-        lines = []
+        lines = ['Basic listing:']
```

Comment:
Leave out this heading. I think the difference between a BASIC listing and the variables is clear enough.